### PR TITLE
Initial support for enabling MyPy in CI 

### PR DESCRIPTION
**Pull request by mikepapadim**
This PR adds initial support for `MyPy` in the Jenkins pipeline.
At this point, it checks only `python/tvm/runtime` and `python/tvm/auto_scheduler`.

1) It adds check with `MyPy` in linting stage,
2) It add an option in `Makefile` for `$ make mypy`.

@jroesch @jwfromm

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks @mikepapadim ! Can we merge the mypy check to the ci lint stage? This is mainly to simplify the pipeline. We can also push the items to https://github.com/apache/tvm/blob/main/tests/scripts/task_lint.sh

**Comment by mikepapadim** [comment]
> Thanks @mikepapadim ! Can we merge the mypy check to the ci lint stage? This is mainly to simplify the pipeline. We can also push the items to https://github.com/apache/tvm/blob/main/tests/scripts/task_lint.sh

Sure, I will revert the changes and keep it in the lint stage.

File: tests/scripts/task_mypy.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+echo "Install all missing stub packages"
+mypy --install-types
+
+echo "Checking MyPy Type defs in auto_scheduler package."
+mypy  --check-untyped-defs python/tvm/auto_scheduler
+
+echo "Checking MyPy Type defs in runtime package."
+mypy  --check-untyped-defs python/tvm/runtime 
+
```
Comment (tqchen) [inline tests/scripts/task_mypy.sh:29]: remove new lines here.

**Comment by jcf94** [comment]
@mikepapadim Thanks! It's really great to have type check for python codes. While seems the CI failed on mypy but it didn't throw any error?

In my understanding, mypy requires python files to specifie data type explicitly for function parameters and return values, but we have not yet add the missing data type to these two dirs in this PR?

**Comment by mikepapadim** [comment]
> @mikepapadim Thanks! It's really great to have type check for python codes. While seems the CI failed on mypy but it didn't throw any error?
> 
> In my understanding, mypy requires python files to specifie data type explicitly for function parameters and return values, but we have not yet add the missing data type to these two dirs in this PR?

@jcf94  you are right. This PR does not fix the missing data types, it is only enables MyPy.

 At this point error logging is disabled ( by setting `ignore_errors` in `mypy.ini`). So, thats why at the end of the linting stage in Jenkins  there is the following:
```
Checking MyPy Type defs in auto_scheduler package.

Success: no issues found in 17 source files

Checking MyPy Type defs in runtime package.

Success: no issues found in 13 source files
```
These two dirs are serving as placeholders atm.

Ideally, this PR adds the basic functionality. 
Then, subsequent PRs can start incrementally to fix data types in specifics  dirs and enable them.

**Comment by jcf94** [comment]
@mikepapadim Thanks, that make sense to me. Looking forward to see all these python files have been filled with data types.

p.s. I just find a:
```
mypy: error: unrecognized arguments: --install-types
```
error in the CI log, does that need to be solved?

**Comment by mikepapadim** [comment]
> @mikepapadim Thanks, that make sense to me.
> 
> p.s. I just find a:
> 
> ```
> mypy: error: unrecognized arguments: --install-types
> ```
> 
> error in the CI log, does that need to be solved?
I been looking into this. 
It looks like MyPy version issue. I will try to reproduce it locally.
https://github.com/python/mypy/issues/10596

**Comment by comaniac** [comment]
It's better to specify all linting/format checking tool versions (i.e., pylint, mypy, black, clang-format) to make sure the results are consistent.

**Review by tqchen — APPROVED** [review]

**Review by jcf94 — APPROVED** [review]

**Comment by junrushao1994** [comment]
Finally!! Thanks for the great work!

As TensorIR is fully prepared for type annotations, we should probably check `python/tvm/tir/schedule` as well. CC: @Hzfengsy @tqchen

**Comment by tqchen** [comment]
@junrushao1994 we can followup with a PR

**Comment by mikepapadim** [comment]
> Finally!! Thanks for the great work!
> 
> As TensorIR is fully prepared for type annotations, we should probably check `python/tvm/tir/schedule` as well. CC: @Hzfengsy @tqchen

@junrushao1994 https://github.com/apache/tvm/pull/8367 Please have a look in this one

**Pull request merged**